pkg/leeway/reporter.go

```diff
@@ -225,3 +228,148 @@ func (r *WerftReporter) PackageBuildFinished(pkg *Package, err error) {
 	}
 	fmt.Printf("[%s|%s] %s\n", pkg.FullName(), status, msg)
 }
+
+type PackageReport struct {
+	logs     strings.Builder
+	start    time.Time
+	duration time.Duration
+	status   PackageBuildStatus
+	err      error
+}
+
+func (r *PackageReport) StatusIcon() string {
+	if r.HasError() {
+		return "❌"
+	}
+	switch r.status {
+	case PackageBuilt:
+		return "✅"
+	case PackageBuilding:
+		return "🏃"
+	case PackageNotBuiltYet:
+		return "🔧"
+	default:
+		return "?"
+	}
+}
+
+func (r *PackageReport) HasError() bool {
+	return r.err != nil
+}
+
+func (r *PackageReport) DurationInSeconds() string {
+	return fmt.Sprintf("%.2fs", r.duration.Seconds())
+}
+
+func (r *PackageReport) Logs() string {
+	return strings.TrimSpace(r.logs.String())
+}
+
+func (r *PackageReport) Error() string {
+	return fmt.Sprintf("%s", r.err)
+}
+
+type MarkdownReporter struct {
+	delegate    Reporter
+	filename    string
+	reports     map[string]*PackageReport
+	rootPackage *Package
+	mu          sync.RWMutex
+}
+
+func NewMarkdownReporter(del Reporter, filename string) *MarkdownReporter {
+	return &MarkdownReporter{
+		delegate: del,
+		filename: filename,
+		reports:  make(map[string]*PackageReport),
+	}
+}
+
+func (r *MarkdownReporter) getReport(pkg *Package) *PackageReport {
+	name := pkg.FullName()
+
+	r.mu.RLock()
+	rep, ok := r.reports[name]
+	r.mu.RUnlock()
+
+	if !ok {
+		r.mu.Lock()
+		rep, ok = r.reports[name]
+		if ok {
+			r.mu.Unlock()
+			return rep
+		}
+
+		rep = &PackageReport{status: PackageNotBuiltYet}
+		r.reports[name] = rep
+		r.mu.Unlock()
+	}
+
+	return rep
+}
+
+func (r *MarkdownReporter) BuildStarted(pkg *Package, status map[*Package]PackageBuildStatus) {
+	r.rootPackage = pkg
+	r.delegate.BuildStarted(pkg, status)
+}
+
+func (r *MarkdownReporter) BuildFinished(pkg *Package, err error) {
+	r.delegate.BuildFinished(pkg, err)
+	r.Report()
+}
+
+func (r *MarkdownReporter) PackageBuildStarted(pkg *Package) {
+	r.delegate.PackageBuildStarted(pkg)
+	rep := r.getReport(pkg)
+	rep.start = time.Now()
+	rep.status = PackageBuilding
+}
+
+func (r *MarkdownReporter) PackageBuildLog(pkg *Package, isErr bool, buf []byte) {
+	report := r.getReport(pkg)
+	report.logs.Write(buf)
+	r.delegate.PackageBuildLog(pkg, isErr, buf)
+}
+
+func (r *MarkdownReporter) PackageBuildFinished(pkg *Package, err error) {
+	r.delegate.PackageBuildFinished(pkg, err)
+	rep := r.getReport(pkg)
+	rep.duration = time.Since(rep.start)
+	rep.status = PackageBuilt
+	rep.err = err
+}
+
+func (r *MarkdownReporter) Report() {
+	vars := make(map[string]interface{})
+	vars["Name"] = r.filename
+	vars["Packages"] = r.reports
+	vars["RootPackage"] = r.rootPackage
+
+	tmplString := `
```

Comment:
I'd prefer to just emit HTML instead of markdown. The template is already heavily using HTML and it would mean we don't have to do the string replace of <code>\`</code> and `'` - we still have to stick to the subset of HTML that markdown supports of course, but I think it's easier to read the template string we just use plain HTML. So replace `#` with `h1`,  <code>\`</code> with `<code>` and so on.

Reply:
I checked if the Markdown renderer also renders pure HTML - and it does. Thus I've changed this PR to generate HTML.